File: 0077-combinations/0077-combinations.cpp
class Solution {
private:
    void backtrack(int start, int n, int k, vector<int>& curr, vector<vector<int>>& ans){
        if (curr.size() == k) {
            ans.push_back(curr);
            return;
        }
        
        for (int i = start; i <= n; ++i){
            curr.push_back(i);
            backtrack(i + 1, n, k, curr, ans);
            curr.pop_back();
        }
    }
    
public:
    vector<vector<int>> combine(int n, int k) {
        vector<vector<int>> ans;
        vector<int> curr;
        backtrack(1, n, k, curr, ans);
        
        return ans;
    }
};

File: 0077-combinations/0077-combinations.py
class Solution:
    def backtrack(self, start: int, n: int, k: int, curr: List[int], ans: List[List[int]]) -> None:
        if len(curr) == k:
            ans.append(curr[:])
            return
        
        for i in range(start, n + 1):
            curr.append(i)
            self.backtrack(i + 1, n, k, curr, ans)
            curr.pop()
        
    
    def combine(self, n: int, k: int) -> List[List[int]]:
        ans = []
        self.backtrack(1, n, k, [], ans)
        
        return ans